File: apps/server-go/internal/game/registry.go
package game

import (
	"sync"
	"time"
)

type Registry struct {
	mu          sync.RWMutex
	games       map[string]*Actor
	stopCh      chan struct{}
	onTickerLag func(lagMS int64)
}

func NewRegistry() *Registry {
	return &Registry{games: make(map[string]*Actor)}
}

func (r *Registry) Register(actor *Actor) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.games[actor.ID] = actor
}

func (r *Registry) Unregister(gameID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.games, gameID)
}

func (r *Registry) Get(gameID string) *Actor {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.games[gameID]
}

func (r *Registry) All() []*Actor {
	r.mu.RLock()
	defer r.mu.RUnlock()
	games := make([]*Actor, 0, len(r.games))
	for _, actor := range r.games {
		games = append(games, actor)
	}
	return games
}

func (r *Registry) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.games)
}

func (r *Registry) SetTickerLagHandler(fn func(lagMS int64)) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.onTickerLag = fn
}

func (r *Registry) StartTicker() {
	r.mu.Lock()
	if r.stopCh != nil {
		r.mu.Unlock()
		return
	}
	stopCh := make(chan struct{})
	r.stopCh = stopCh
	r.mu.Unlock()

	go func() {
		interval := time.Second
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		expected := time.Now().Add(interval)
		for {
			select {
			case fired := <-ticker.C:
				lag := fired.Sub(expected).Milliseconds()
				if lag < 0 {
					lag = 0
				}
				r.mu.RLock()
				handler := r.onTickerLag
				r.mu.RUnlock()
				if handler != nil {
					handler(lag)
				}
				for _, actor := range r.All() {
					actor.Tick(fired)
				}
				expected = fired.Add(interval)
			case <-stopCh:
				return
			}
		}
	}()
}

func (r *Registry) StopTicker() {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.stopCh == nil {
		return
	}
	close(r.stopCh)
	r.stopCh = nil
}

File: apps/server/src/game/registry.ts
import { logger } from "../logger.ts";
import type { GameActor } from "./game-actor.ts";

const TICK_INTERVAL_MS = 1000;

class GameRegistry {
  private readonly games: Map<string, GameActor> = new Map();
  private ticker: ReturnType<typeof setInterval> | null = null;

  register(game: GameActor): void {
    this.games.set(game.id, game);
  }

  unregister(gameId: string): void {
    this.games.delete(gameId);
  }

  get(gameId: string): GameActor | undefined {
    return this.games.get(gameId);
  }

  all(): GameActor[] {
    return Array.from(this.games.values());
  }

  startTicker(): void {
    if (this.ticker !== null) return;
    this.ticker = setInterval(() => {
      const now = Date.now();
      for (const game of this.games.values()) {
        try {
          game.tick(now);
        } catch (err) {
          logger.error("game tick failed", { gameId: game.id, err: String(err) });
        }
      }
    }, TICK_INTERVAL_MS);
    if (typeof this.ticker.unref === "function") this.ticker.unref();
  }

  stopTicker(): void {
    if (this.ticker !== null) {
      clearInterval(this.ticker);
      this.ticker = null;
    }
  }
}

export const registry = new GameRegistry();

export function startTicker(): void {
  registry.startTicker();
}
